File: ios/SampleUI/Tests/SampleUITests/UseSmileIDSampleProfilesTest.swift
@testable import SampleUI
import XCTest

final class UseSmileIDSampleProfilesTest: XCTestCase {
  private let ada = UseSmileIDSampleUserDetails(firstName: "Ada", lastName: "Okafor", email: "ada@kobo.example")

  func testInitialsComeFromThePersonThenTheOrganisation() {
    XCTAssertEqual(profile(first: "Kwame", last: "Asante").initials, "KA")
    XCTAssertEqual(profile(first: "Mary Anne", last: "Smith").initials, "MA")
    XCTAssertEqual(profile(first: "amina").initials, "A")
    XCTAssertEqual(profile(organisation: "PesaLink").initials, "P")
    XCTAssertEqual(profile(organisation: "").initials, "?")
  }

  func testAPlainLaunchHasNoProfile() {
    let profiles = UseSmileIDSampleProfiles.forLaunch(seedProfiles: false, stored: UseSmileIDSampleProfiles())

    XCTAssertTrue(profiles.all.isEmpty)
    XCTAssertNil(profiles.active)
    XCTAssertEqual(profiles.partnerName, "Smile ID")
    XCTAssertEqual(profiles.partnerId, "p-1")
  }

  func testASeededLaunchShowsTheFixturesOverWhatWasStored() {
    let stored = UseSmileIDSampleProfiles([profile(organisation: "Kobo Bank")])

    XCTAssertEqual(UseSmileIDSampleProfiles.forLaunch(seedProfiles: true, stored: stored).all, UseSmileIDSampleProfiles.fixtures())
    XCTAssertEqual(UseSmileIDSampleProfiles.forLaunch(seedProfiles: false, stored: stored), stored)
  }

  func testTheFirstProfileBecomesActiveAndLaterOnesWaitForTheOffer() {
    var profiles = UseSmileIDSampleProfiles()

    let first = profiles.add(organisation: "Karibu Pay")
    XCTAssertEqual(profiles.activeId, first.id)
    XCTAssertNil(profiles.lastCreatedId, "the active one needs no Make active offer")

    let second = profiles.add(organisation: "Sahara Pay")
    XCTAssertEqual(profiles.activeId, first.id)
    XCTAssertEqual(profiles.lastCreatedId, second.id)
    XCTAssertEqual(profiles.all.map(\.id), ["p-1", "p-2"])
  }

  func testIdsSkipOnesAlreadyTaken() {
    var profiles = UseSmileIDSampleProfiles([profile(id: "p-1"), profile(id: "p-3"), profile(id: "p-4")])

    XCTAssertEqual(profiles.add(organisation: "Acme").id, "p-5")
    XCTAssertEqual(profiles.add(organisation: "Beta").id, "p-6")
  }

  func testABlankOrganisationNamesTheAppOnConsentNeverThePerson() {
    var profiles = UseSmileIDSampleProfiles()
    profiles.add(organisation: "", defaults: ada)

    XCTAssertEqual(profiles.active?.title, "Ada Okafor")
    XCTAssertEqual(profiles.partnerName, "Smile ID")
  }

  func testAnUpdateLeavesWhatItWasNotGiven() {
    var profiles = UseSmileIDSampleProfiles()
    let created = profiles.add(organisation: "Karibu Pay")
    profiles.update(created.id, callbackUrl: " https://partner.example/hook ")

    profiles.update(created.id, defaults: ada)

    XCTAssertEqual(profiles.active?.callbackUrl, "https://partner.example/hook")
    XCTAssertEqual(profiles.active?.organisation, "Karibu Pay")
    XCTAssertEqual(profiles.active?.person, "Ada Okafor")
  }

  func testDeletingTheActiveProfileHandsOverToTheFirstLeftThenToNone() {
    var profiles = UseSmileIDSampleProfiles(UseSmileIDSampleProfiles.fixtures(), activeId: "p-2")

    profiles.delete("p-2")
    XCTAssertEqual(profiles.activeId, "p-1")

    profiles.delete("p-1")
    profiles.delete("p-3")
    XCTAssertNil(profiles.activeId)
  }

  func testSignOutClearsEveryProfile() {
    var profiles = UseSmileIDSampleProfiles(UseSmileIDSampleProfiles.fixtures())

    profiles.clear()

    XCTAssertEqual(profiles, UseSmileIDSampleProfiles())
  }

  func testAStoredActiveIdThatNamesNoProfileFallsBackToTheFirst() {
    XCTAssertEqual(UseSmileIDSampleProfiles(UseSmileIDSampleProfiles.fixtures(), activeId: "p-9").activeId, "p-1")
  }

  func testTheFirstRunCreatesAnActiveProfileFromWhatWasTyped() {
    var profiles = UseSmileIDSampleProfiles()

    profiles.keep(ada, organisation: " Kobo Bank ")

    XCTAssertEqual(profiles.active?.organisation, "Kobo Bank")
    XCTAssertEqual(profiles.active?.defaults, ada)
  }

  func testAnActiveProfileTakesTheEdits() {
    var profiles = UseSmileIDSampleProfiles(UseSmileIDSampleProfiles.fixtures())

    profiles.keep(ada, organisation: "")

    XCTAssertEqual(profiles.find("p-1")?.defaults, ada)
    XCTAssertEqual(profiles.find("p-1")?.organisation, "UpTech Finance")
    XCTAssertEqual(profiles.all.count, 3)
  }

  func testAFieldTheTokenSuppliesIsNeverStored() {
    let stored = UseSmileIDSampleUserDetails(firstName: "Kwame", lastName: "Asante")
    var profiles = UseSmileIDSampleProfiles([UseSmileIDSampleProfile(id: "p-1", organisation: "UpTech", defaults: stored)])
    let namesBound = UseSmileIDSampleUserDetailsRequirement(
      bindings: UseSmileIDSampleTokenBindings(givenNames: true, lastName: true)
    )

    profiles.keep(UseSmileIDSampleUserDetails(email: "ada@kobo.example"), organisation: "", requirement: namesBound)

    XCTAssertEqual(profiles.active?.defaults, UseSmileIDSampleUserDetails(firstName: "Kwame", lastName: "Asante", email: "ada@kobo.example"))
  }

  func testARecordRoundTripsIncludingCharactersJsonEscapes() {
    let profiles = UseSmileIDSampleProfiles(
      [
        UseSmileIDSampleProfile(
          id: "p-1",
          organisation: "Kobo \"Bank\" \\ Ltd\n",
          defaults: UseSmileIDSampleUserDetails(firstName: "Adá", lastName: "O'Neil", email: "ada@kobo.example", phone: "+254 700"),
          callbackUrl: "https://kobo.example/hook?a=1&b=2"
        ),
        UseSmileIDSampleProfile(id: "p-2", organisation: "")
      ],
      activeId: "p-2"
    )

    XCTAssertEqual(UseSmileIDSampleProfilesCodec.decode(UseSmileIDSampleProfilesCodec.encode(profiles)), profiles)
  }

  func testUnreadableInputIsNoProfiles() {
    let inputs: [String?] = [
      nil,
      "",
      "not json",
      "[]",
      #"{"profiles":[{"id":"p-1"}]}"#,
      #"{"version":2,"profiles":[{"id":"p-1"}]}"#,
      #"{"version":1,"profiles":"p-1"}"#,
      #"{"version":1,"profiles":[{"id":"p-1","organisation":"Kobo""#
    ]
    for text in inputs {
      XCTAssertEqual(UseSmileIDSampleProfilesCodec.decode(text.map { Data($0.utf8) }), UseSmileIDSampleProfiles(), text ?? "nil")
    }
  }

  func testAProfileWithoutAnIdOrWithARepeatedOneIsDropped() {
    let text = #"{"version":1,"activeId":"p-1","profiles":[{"organisation":"No id"},{"id":"p-1","organisation":"First"},{"id":"p-1","organisation":"Again"}]}"#

    XCTAssertEqual(UseSmileIDSampleProfilesCodec.decode(Data(text.utf8)).all.map(\.organisation), ["First"])
  }

  func testTheAndroidShapeReadsTheSame() {
    let android = #"{"version":1,"activeId":"p-1","profiles":[{"id":"p-1","organisation":"Kobo","firstName":"Ada","lastName":"","email":"","phone":"","callbackUrl":""}]}"#

    let profiles = UseSmileIDSampleProfilesCodec.decode(Data(android.utf8))

    XCTAssertEqual(profiles.active?.organisation, "Kobo")
    XCTAssertEqual(profiles.active?.defaults.firstName, "Ada")
  }

  private func profile(
    id: String = "p-1",
    organisation: String = "UpTech Finance",
    first: String = "",
    last: String = ""
  ) -> UseSmileIDSampleProfile {
    UseSmileIDSampleProfile(
      id: id,
      organisation: organisation,
      defaults: UseSmileIDSampleUserDetails(firstName: first, lastName: last)
    )
  }
}

File: flutter/sample_ui/test/state/use_smileid_sample_profiles_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:sample_ui/sample_ui.dart';

void main() {
  const UseSmileIDSampleUserDetails ada = UseSmileIDSampleUserDetails(
    firstName: 'Ada',
    lastName: 'Okafor',
    email: 'ada@kobo.example',
  );

  UseSmileIDSampleProfile profile({
    String id = 'p-1',
    String organisation = 'UpTech Finance',
    String first = '',
    String last = '',
  }) => UseSmileIDSampleProfile(
    id: id,
    organisation: organisation,
    defaults: UseSmileIDSampleUserDetails(firstName: first, lastName: last),
  );

  test('a plain launch has no profile', () {
    final UseSmileIDSampleProfiles profiles =
        UseSmileIDSampleProfiles.forLaunch(
          const UseSmileIDSampleLaunchArgs(),
          stored: UseSmileIDSampleProfiles(),
        );

    expect(profiles.all, isEmpty);
    expect(profiles.active, isNull);
    expect(profiles.partnerName, 'Smile ID');
    expect(profiles.partnerId, 'p-1');
  });

  test('a seeded launch shows the fixtures over what was stored', () {
    final UseSmileIDSampleProfiles stored = UseSmileIDSampleProfiles(
      <UseSmileIDSampleProfile>[profile(organisation: 'Kobo Bank')],
    );

    expect(
      UseSmileIDSampleProfiles.forLaunch(
        const UseSmileIDSampleLaunchArgs(seedProfiles: true),
        stored: stored,
      ).all,
      UseSmileIDSampleProfiles.fixtures(),
    );
    expect(
      UseSmileIDSampleProfiles.forLaunch(
        const UseSmileIDSampleLaunchArgs(),
        stored: stored,
      ),
      same(stored),
    );
  });

  group('initials', () {
    test('come from the person, then the organisation', () {
      expect(profile(first: 'Kwame', last: 'Asante').initials, 'KA');
      expect(profile(first: 'Mary Anne', last: 'Smith').initials, 'MA');
      expect(profile(organisation: 'PesaLink').initials, 'P');
      expect(profile(organisation: '').initials, '?');
    });
  });

  test(
    'the first profile becomes active and later ones wait for the offer',
    () {
      final UseSmileIDSampleProfiles profiles = UseSmileIDSampleProfiles();

      final UseSmileIDSampleProfile first = profiles.add(
        organisation: 'Karibu Pay',
      );
      expect(profiles.activeId, first.id);
      expect(profiles.lastCreatedId, isNull);

      final UseSmileIDSampleProfile second = profiles.add(
        organisation: 'Sahara Pay',
      );
      expect(profiles.activeId, first.id);
      expect(profiles.lastCreatedId, second.id);
      expect(profiles.all.map((UseSmileIDSampleProfile p) => p.id), <String>[
        'p-1',
        'p-2',
      ]);
    },
  );

  test('ids skip ones already taken', () {
    final UseSmileIDSampleProfiles profiles = UseSmileIDSampleProfiles(
      <UseSmileIDSampleProfile>[
        profile(),
        profile(id: 'p-3'),
        profile(id: 'p-4'),
      ],
    );

    expect(profiles.add(organisation: 'Acme').id, 'p-5');
    expect(profiles.add(organisation: 'Beta').id, 'p-6');
  });

  test('a blank organisation names the app on consent, never the person', () {
    final UseSmileIDSampleProfiles profiles = UseSmileIDSampleProfiles()
      ..add(organisation: '', defaults: ada);

    expect(profiles.active?.title, 'Ada Okafor');
    expect(profiles.partnerName, 'Smile ID');
  });

  test('an update leaves what it was not given', () {
    final UseSmileIDSampleProfiles profiles = UseSmileIDSampleProfiles();
    final UseSmileIDSampleProfile created = profiles.add(
      organisation: 'Karibu Pay',
    );
    profiles.update(created.id, callbackUrl: ' https://partner.example/hook ');

    profiles.update(created.id, defaults: ada);

    expect(profiles.active?.callbackUrl, 'https://partner.example/hook');
    expect(profiles.active?.organisation, 'Karibu Pay');
    expect(profiles.active?.person, 'Ada Okafor');
  });

  test(
    'deleting the active profile hands over to the first left, then to none',
    () {
      final UseSmileIDSampleProfiles profiles = UseSmileIDSampleProfiles(
        UseSmileIDSampleProfiles.fixtures(),
        'p-2',
      );

      profiles.delete('p-2');
      expect(profiles.activeId, 'p-1');

      profiles
        ..delete('p-1')
        ..delete('p-3');
      expect(profiles.activeId, isNull);
    },
  );

  test('sign out clears every profile', () {
    final UseSmileIDSampleProfiles profiles = UseSmileIDSampleProfiles(
      UseSmileIDSampleProfiles.fixtures(),
    )..clear();

    expect(profiles.all, isEmpty);
    expect(profiles.active, isNull);
  });

  test('a stored active id that names no profile falls back to the first', () {
    expect(
      UseSmileIDSampleProfiles(
        UseSmileIDSampleProfiles.fixtures(),
        'p-9',
      ).activeId,
      'p-1',
    );
  });

  group('keep', () {
    test('the first run creates an active profile from what was typed', () {
      final UseSmileIDSampleProfiles profiles = UseSmileIDSampleProfiles()
        ..keep(ada, organisation: ' Kobo Bank ');

      expect(profiles.active?.organisation, 'Kobo Bank');
      expect(profiles.active?.defaults, ada);
    });

    test('an active profile takes the edits', () {
      final UseSmileIDSampleProfiles profiles = UseSmileIDSampleProfiles(
        UseSmileIDSampleProfiles.fixtures(),
      )..keep(ada, organisation: '');

      expect(profiles.find('p-1')?.defaults, ada);
      expect(profiles.find('p-1')?.organisation, 'UpTech Finance');
      expect(profiles.all, hasLength(3));
    });

    test('a field the token supplies is never stored', () {
      final UseSmileIDSampleProfiles profiles =
          UseSmileIDSampleProfiles(<UseSmileIDSampleProfile>[
            profile(organisation: 'UpTech', first: 'Kwame', last: 'Asante'),
          ])..keep(
            const UseSmileIDSampleUserDetails(email: 'ada@kobo.example'),
            organisation: '',
            requirement: const UseSmileIDSampleUserDetailsRequirement(
              firstName: false,
              lastName: false,
            ),
          );

      expect(
        profiles.active?.defaults,
        const UseSmileIDSampleUserDetails(
          firstName: 'Kwame',
          lastName: 'Asante',
          email: 'ada@kobo.example',
        ),
      );
    });
  });

  group('the stored record', () {
    test('round-trips, including characters JSON escapes', () {
      final UseSmileIDSampleProfiles profiles =
          UseSmileIDSampleProfiles(<UseSmileIDSampleProfile>[
            const UseSmileIDSampleProfile(
              id: 'p-1',
              organisation: 'Kobo "Bank" \\ Ltd\n',
              defaults: UseSmileIDSampleUserDetails(
                firstName: 'Adá',
                lastName: "O'Neil",
                email: 'ada@kobo.example',
                phone: '+254 700',
              ),
              callbackUrl: 'https://kobo.example/hook?a=1&b=2',
            ),
            const UseSmileIDSampleProfile(id: 'p-2', organisation: ''),
          ], 'p-2');

      final UseSmileIDSampleProfiles decoded =
          UseSmileIDSampleProfilesCodec.decode(
            UseSmileIDSampleProfilesCodec.encode(profiles),
          );

      expect(decoded.all, profiles.all);
      expect(decoded.activeId, 'p-2');
    });

    test('anything unreadable is no profiles', () {
      for (final String? text in <String?>[
        null,
        '',
        'not json',
        '[]',
        '{"profiles":[{"id":"p-1"}]}',
        '{"version":2,"profiles":[{"id":"p-1"}]}',
        '{"version":1,"profiles":"p-1"}',
        '{"version":1,"profiles":[{"id":"p-1","organisation":"Kobo"',
      ]) {
        expect(
          UseSmileIDSampleProfilesCodec.decode(text).all,
          isEmpty,
          reason: text,
        );
      }
    });

    test('a profile without an id, or with a repeated one, is dropped', () {
      final UseSmileIDSampleProfiles
      decoded = UseSmileIDSampleProfilesCodec.decode(
        '{"version":1,"activeId":"p-1","profiles":[{"organisation":"No id"},'
        '{"id":"p-1","organisation":"First"},{"id":"p-1","organisation":"Again"}]}',
      );

      expect(
        decoded.all.map((UseSmileIDSampleProfile p) => p.organisation),
        <String>['First'],
      );
    });

    test('reads the shape the other three apps write', () {
      final UseSmileIDSampleProfiles
      decoded = UseSmileIDSampleProfilesCodec.decode(
        '{"version":1,"activeId":"p-1","profiles":[{"id":"p-1","organisation":"Kobo",'
        '"firstName":"Ada","lastName":"","email":"","phone":"","callbackUrl":""}]}',
      );

      expect(decoded.active?.organisation, 'Kobo');
      expect(decoded.active?.defaults.firstName, 'Ada');
    });
  });
}
